New generator-iterator tag (#2243)
python-discord/bot: bot/resources/tags/generators-iterators.md

**Generators and Iterators**

What is a generator and an iterator and what is the difference between each other?

**Generator**

> A generator is a function that may contain a `yield` or sometimes `return` statements, a generator can also be an expression(genexpr).

Reviewer: This seems problematic, generators don't have to be functions. You can create a generator class by implementing methods defined in `collections.abc.Generator`

```py
>>> print(num for num in range(11))
<generator object <genexpr> at 0x0000000000000000>
```

Well, what does a generator do and what is it any different from a list, for example?
The generator yields one item at a time and generates items only when in demand. Whereas, in a list, Python reserves memory for the whole list. A generator can be an iterator but not vice versa.

**Iterator**

An iterator is any object whose class has `__next__` and `__iter__` methods. An iterator can also be created with the built-in function `iter`

```py
from typing import NoReturn

class Foo:
    def __init__(self, max: int = 0) -> None:
        self.max = max

    def __iter__(self) -> "Foo":
        self.n = 0
        return self

    def __next__(self) -> int | NoReturn:
        if self.n <= self.max:
            result = 2**self.n
            self.n += 1
            return result
        else:
            raise StopIteration
```
The `StopIteration` exception must be raised or the iterator is pronounced broken.

**What is the difference between iterators and generators?**

> An iterator requires a `__next__` and `__iter__` methods over a generator that only needs an `__iter__` method.

Reviewer: This is incorrect. Merely implementing `__iter__` does not create a `Generator`, you can test it as such:

```py
import typing

class Foo:
    def __iter__(self):
        return self

print(isinstance(Foo(), typing.Generator))
>> False
```

All Generators are Iterators, the abstract base class for a `Generator` actually inherits `Iterator`. `Generator` actually requires 2 methods to be implemented over an `Iterator`'s `__next__` and `__iter__`, which are `send`, `throw`, and the mix-in methods `close`

More info here: https://docs.python.org/3/library/collections.abc.html

---
aliases: ["generators", "iterators", "yield"]
embed:
    title: "Generators and Iterators"
---

[Iterators](https://docs.python.org/3/glossary.html#term-iterator) are objects that support iteration. Specifically, they need to implement the `__iter__` and `__next__` special methods and raise `StopIteration` when there are no values to be returned.

Built-in functions such as [zip](https://docs.python.org/3/library/functions.html#zip) and [map](https://docs.python.org/3/library/functions.html#map) return Iterators, this allows lazy, or memory-efficient, iteration as the elements do not need to be pre-computed before being used.

[Generators](https://docs.python.org/3/glossary.html#term-generator) are extended Iterators, which provide a simple way to create objects that function like Iterators by using yield functions or expressions.

**Generator Functions**

A generator function can be created by using one or more `yield` statements.
\```py
def foo():
    yield 1
    yield 2
\```

> When called, this function returns a Generator. Code within the generator does not execute immediately when a Generator is created.

Reviewer: ```suggestion
Generators can be a memory-efficient alternative to lists in many situations.
```

> \```py
> foo()
> <generator object foo at 0x1>
> \```
> Generator results can be accessed by iteration with `for`, or using `next()`. When calling `next()` on a Generator, the function will run until it reaches the next yield, then pause and yield the result. Further code does not execute until the next `yield` is reached. If the Generator is exhausted, a `StopIteration` Exception is raised.
> \```py
> x = foo()
> next(x) -> 1
> next(x) -> 2
> next(x) -> StopIteration
> \```

Reviewer: ```suggestion
```

Removed detailed discussion of generator states.

**Generator Expressions**

> Generators can also be created by `for` clauses inside [expressions](https://docs.python.org/3/glossary.html#term-generator-expression). This is similar to creating a list comprehension but using `()` instead of `[]`. You can use `if` `else` ternary operators as you would in other comprehension forms.

Reviewer: ```suggestion
Generators can also be created by `for` clauses inside [expressions](https://docs.python.org/3/glossary.html#term-generator-expression), similar to creating a list comprehension.
```
Made expression introduction more concise.

\```py

> foo = (x * 2 for x in range(5))
> bar = sum(x * 2 for x in range(5))

Reviewer: ```suggestion
foo = (x * 2 for x in range(5))

for num in foo:
    print(num)
```

Instead of using sum, used print example which might be easier to understand.

\```

> Note that Generators created using expressions are not functions that return Generators, rather, they are already Generator objects. You can only iterate the contents of `foo` once before it is exhausted.
> \```py
> list(foo) -> [0, 2, 4, 6, 8]
> list(foo) -> []
> \```
> 
> You can also use `*` to unpack Generators as positional arguments for functions, like any other Iterable object.
> \```py
> print(*foo) -> 0 2 4 6 8
> \```

Reviewer: ```suggestion
```

Removed extra details on generator states and unpacking

**More Information**

> • Generators and Iterators can also be created as custom classes by implementing their required [methods](https://docs.python.org/3/library/collections.abc.html#collections-abstract-base-classes-1). The preferred way to do this is by extending the respective Abstract Base Classes within `collections.abc`.

Reviewer: ```suggestion
• Generators and Iterators can also be created as custom classes by implementing their required [abstract methods](https://docs.python.org/3/library/collections.abc.html#collections-abstract-base-classes-1).
```

Made ABC link more concise

What is a generator and an iterator and what is the difference between each other?

**Generator**

A generator yields one item at a time and generates items only when in demand and can be created with a function that may contain a `yield` or sometimes `return` statements, but they're not bound to functions as you can subclass [collections.abc.Generator](https://docs.python.org/3/library/collections.abc.html#collections.abc.Generator). Here we create an infinite generator that can be an iterator.
```py
from typing import Generator, Iterator

def infinite_stream(start: int) -> Generator[int, None, None] | Iterator[int]:
    while True:
        yield start
        start += 1
```

A generator can also be an expression(genexpr).

```py
>>> print(num for num in range(11))
<generator object <genexpr> at 0x0000000000000000>
```

How is a generator any different from a list, for example?
The generator yields one item at a time and generates items only when in demand. Whereas, in a list, Python reserves memory for the whole list.

**Iterator**

An iterator is any object whose class has `__next__` and `__iter__` methods. An iterator can also be created with the built-in function `iter`

> ```py
> from typing import NoReturn
> 
> 
> class Foo:
>     def __init__(self, max: int = 0) -> None:
>         self.max = max
> 
>     def __iter__(self) -> "Foo":
>         self.n = 0
>         return self
> 
>     def __next__(self) -> int | NoReturn:
>         if self.n <= self.max:
>             result = 2**self.n
>             self.n += 1
>             return result
>         else:
>             raise StopIteration
> ```
> The `StopIteration` exception must be raised or the iterator is pronounced broken.
> 

Reviewer: ```suggestion
\```py
foo = (x * 2 for x in range(5))
bar = sum(x * 2 for x in range(5))
\```

Note that Generators created using expressions are not functions that return Generators, rather, they are already Generator objects. You can only iterate the contents of `foo` once before it is exhausted.
\```py
list(foo) -> [0, 2, 4, 6, 8]
list(foo) -> []
\```

You can also use `*` to unpack Generators as positional arguments for functions, like any other Iterable object.
\```py
print(*foo) -> 0 2 4 6 8
\```
```
Rationale explained in previous suggestion, here it might be a better use of content to explain some examples on generator and expression usage instead of building custom Iterator classes.

> ### Note: Please unescape the `\` marks before each of the code blocks, I can't find any other way to nest codeblocks inside the suggestion codeblock.

> **What is the difference between iterators and generators?**
> 
> All Generators are Iterators, the abstract base class for a `Generator` actually inherits `Iterator`. `Generator` actually requires 2 methods to be implemented over an `Iterator`'s `__next__` and `__iter__`, which are `send`, `throw`, and the mix-in methods `close`, the only difference of a `Generator` and an `Iterator` are there use cases and implementation.
> 
> More info here: [collections.abc](https://docs.python.org/3/library/collections.abc.html)

Reviewer: ```suggestion
**More Information**

• Generators and Iterators can also be created as custom classes by implementing their required [methods](https://docs.python.org/3/library/collections.abc.html#collections-abstract-base-classes-1). The preferred way to do this is by extending the respective Abstract Base Classes within `collections.abc`.

• Check out the [Real python article](https://realpython.com/introduction-to-python-generators/) for a more in-depth view of Generators and their use-cases!
```

If explaining the difference at the start, we don't really need this ending last part, consider changing to an additional resource section.
